File: blender/content/configurationData.py
from enum import Enum
from typing import Any
from mathutils import Vector, Euler

class OutputType(Enum):
    NUMBER = "number"
    VECTOR = "vector"
    EULER = "euler"
    SHAPEKEY = "shapekey"
    BOOLEAN = "boolean"

class InputType(Enum):
    NUMBER = "number"
    VARIABLE = "variable"
    BOOLEAN = "boolean"

class ConfigurationInput:
    id: str
    name: str | None
    type: InputType
    default: Any
    min: float | None
    max: float | None
    comment: str | None
    
    def __init__(self, data):
        self.id = data.get("id", "")
        self.name = data.get("name")
        self.type = InputType(data.get("type", InputType.NUMBER.value))
        self.default = data.get("default")
        self.min = data.get("min")
        self.max = data.get("max")
        self.comment = data.get("comment")
    
class ShapekeyOutput:
    id: str
    input: str
    type: OutputType
    conversion: str
    comment: str | None
    
    def __init__(self, data):
        self.id = data.get("id", "")
        self.input = data.get("input", data.get("inputs", ""))
        self.type = OutputType(data.get("type", OutputType.SHAPEKEY.value))
        self.conversion = data.get("conversion", "")
        self.comment = data.get("comment")
    
class AttachmentPointChild:
    """Describes a child PME wired to a parent attachment point."""
    element_id: str
    target_id: str
    output_value_mapping: dict

    def __init__(self, data):
        self.element_id = data.get("elementId", "")
        self.target_id = data.get("targetId", "")
        mapping = data.get("outputValueMapping", {})
        self.output_value_mapping = mapping if isinstance(mapping, dict) else {}

    def normalized_mapping(self) -> dict:
        """Return mapping with $ prefix ensured on child input IDs."""
        result = {}
        for key, value in self.output_value_mapping.items():
            normalized_key = key if key.startswith("$") else f"${key}"
            result[normalized_key] = value
        return result


class AttachmentPointLocationOutput:
    id: str
    type: OutputType
    input_location: str
    input_rotation: str
    child: "AttachmentPointChild | None"
    comment: str | None
    
    def __init__(self, data):
        self.id = data.get("id", "")
        self.type = OutputType(data.get("type", OutputType.VECTOR.value))
        self.input_location = data.get("inputLocation", data.get("inputs", "vec3(0, 0, 0)"))
        self.input_rotation = data.get("inputRotation", data.get("inputs", "vec3(0, 0, 0)"))
        child_data = data.get("child")
        self.child = AttachmentPointChild(child_data) if isinstance(child_data, dict) else None
        self.comment = data.get("comment")

class ValueOutput:
    id: str
    type: OutputType
    input: str
    comment: str | None
    
    def __init__(self, data):
    
        self.id = data.get("id", "")
        self.type = OutputType(data.get("type", OutputType.NUMBER.value))
        self.input = data.get("input", data.get("inputs", ""))
        self.comment = data.get("comment")
    
class ConfigurationOutput:
    shapekeys: list[ShapekeyOutput] | None
    attachmentpoints: list[AttachmentPointLocationOutput] | None
    values: list[ValueOutput] | None
    
    def __init__(self, data):
        shapekeys = data.get("shapekeys", [])
        attachmentpoints = data.get("attachmentpoints", data.get("attachmentPoints", []))
        values = data.get("values", [])

        self.shapekeys = [ShapekeyOutput(sk) for sk in shapekeys if isinstance(sk, dict)]
        self.attachmentpoints = [AttachmentPointLocationOutput(ap) for ap in attachmentpoints if isinstance(ap, dict)]
        self.values = [ValueOutput(vo) for vo in values if isinstance(vo, dict)]
    
class ConfigurationExpression:
    id: str
    type: OutputType
    inputs: list[str]
    expression: str
    comment: str | None
    
    def __init__(self, data):
        self.id = data.get("id", "")
        self.type = OutputType(data.get("type", OutputType.NUMBER.value))
        inputs = data.get("inputs", [])
        self.inputs = inputs if isinstance(inputs, list) else []
        self.expression = data.get("expression", "")
        self.comment = data.get("comment")
    

class ConfigurationData:
    
    input: list[ConfigurationInput]
    expressions: list[ConfigurationExpression]
    output: ConfigurationOutput
    
    def __init__(self, json_data):
        self.json_data = json_data
        if "input" in json_data and isinstance(json_data["input"], list):
            self.input = [ConfigurationInput(data) for data in json_data["input"] if isinstance(data, dict)]
        else:
            self.input = []
        if "expressions" in json_data and isinstance(json_data["expressions"], list):
            self.expressions = [ConfigurationExpression(data) for data in json_data["expressions"] if isinstance(data, dict)]
        else:
            self.expressions = []
        output_payload = {}
        if "output" in json_data and isinstance(json_data["output"], dict):
            output_payload = dict(json_data["output"])

        if "attachmentpoints" in output_payload:
            pass
        elif "attachmentPoints" in output_payload:
            pass
        elif "attachmentPoints" in json_data and isinstance(json_data["attachmentPoints"], list):
            output_payload["attachmentPoints"] = json_data["attachmentPoints"]

        self.output = ConfigurationOutput(output_payload)
    

File: webapp/src/engine/configurationData.ts
import type {
  AttachmentPointLocationOutput,
  ConfigurationData,
  ConfigurationExpression,
  ConfigurationInput,
  ConfigurationOutput,
  InputType,
  OutputType,
  ShapekeyOutput,
  ValueOutput,
} from './types';

const OUTPUT_TYPES = new Set<OutputType>(['number', 'vector', 'euler', 'shapekey', 'boolean']);
const INPUT_TYPES = new Set<InputType>(['number', 'variable', 'boolean']);

const asObject = (value: unknown): Record<string, unknown> => (value && typeof value === 'object' ? (value as Record<string, unknown>) : {});
const asString = (value: unknown, fallback = ''): string => (typeof value === 'string' ? value : fallback);
const asNumber = (value: unknown): number | undefined => (typeof value === 'number' && Number.isFinite(value) ? value : undefined);
const asStringArray = (value: unknown): string[] => (Array.isArray(value) ? value.filter((item): item is string => typeof item === 'string') : []);

const assertParametricElementType = (root: Record<string, unknown>): void => {
  // In gewone taal: we accepteren alleen JSON's met type "parametricElement".
  if (root.type !== 'parametricElement') {
    throw new Error('Ongeldige PME JSON: alleen type "parametricElement" is toegestaan.');
  }
};

const asOutputType = (value: unknown, fallback: OutputType): OutputType => {
  const candidate = asString(value);
  return OUTPUT_TYPES.has(candidate as OutputType) ? (candidate as OutputType) : fallback;
};

const asInputType = (value: unknown, fallback: InputType): InputType => {
  const candidate = asString(value);
  return INPUT_TYPES.has(candidate as InputType) ? (candidate as InputType) : fallback;
};

const parseInput = (raw: unknown): ConfigurationInput => {
  const data = asObject(raw);
  return {
    id: asString(data.id),
    name: typeof data.name === 'string' ? data.name : undefined,
    type: asInputType(data.type, 'number'),
    default: data.default,
    min: asNumber(data.min),
    max: asNumber(data.max),
    comment: typeof data.comment === 'string' ? data.comment : undefined,
  };
};

const parseExpression = (raw: unknown): ConfigurationExpression => {
  const data = asObject(raw);
  return {
    id: asString(data.id),
    type: asOutputType(data.type, 'number'),
    inputs: asStringArray(data.inputs),
    expression: asString(data.expression),
    comment: typeof data.comment === 'string' ? data.comment : undefined,
  };
};

const parseShapekey = (raw: unknown): ShapekeyOutput => {
  const data = asObject(raw);
  return {
    id: asString(data.id),
    input: asString(data.input, asString(data.inputs)),
    type: asOutputType(data.type, 'shapekey'),
    conversion: asString(data.conversion),
    comment: typeof data.comment === 'string' ? data.comment : undefined,
  };
};

const parseAttachmentPoint = (raw: unknown): AttachmentPointLocationOutput => {
  const data = asObject(raw);
  return {
    id: asString(data.id),
    type: asOutputType(data.type, 'vector'),
    inputLocation: asString(data.inputLocation, asString(data.inputs, 'vec3(0, 0, 0)')),
    inputRotation: asString(data.inputRotation, asString(data.inputs, 'vec3(0, 0, 0)')),
    comment: typeof data.comment === 'string' ? data.comment : undefined,
  };
};

const parseValue = (raw: unknown): ValueOutput => {
  const data = asObject(raw);
  return {
    id: asString(data.id),
    type: asOutputType(data.type, 'number'),
    input: asString(data.input, asString(data.inputs)),
    comment: typeof data.comment === 'string' ? data.comment : undefined,
  };
};

const parseOutput = (rawOutput: unknown, root: Record<string, unknown>): ConfigurationOutput => {
  const output = asObject(rawOutput);
  const attachmentSource = output.attachmentpoints ?? output.attachmentPoints ?? root.attachmentPoints ?? [];

  return {
    shapekeys: (Array.isArray(output.shapekeys) ? output.shapekeys : []).map(parseShapekey),
    attachmentpoints: (Array.isArray(attachmentSource) ? attachmentSource : []).map(parseAttachmentPoint),
    values: (Array.isArray(output.values) ? output.values : []).map(parseValue),
  };
};

export const parseConfigurationData = (payload: unknown): ConfigurationData => {
  const root = asObject(payload);
  assertParametricElementType(root);

  return {
    input: (Array.isArray(root.input) ? root.input : []).map(parseInput),
    expressions: (Array.isArray(root.expressions) ? root.expressions : []).map(parseExpression),
    parts: asStringArray(root.parts),
    output: parseOutput(root.output, root),
  };
};

export const loadConfigurationDataFromJson = (jsonText: string): ConfigurationData => {
  const parsed = JSON.parse(jsonText) as unknown;
  return parseConfigurationData(parsed);
};
